# 3.3 The Geometric Mean
The **geometric mean** is a multiplicative average, as opposed to the arithmetic mean which is an additive average. In general, the geometric mean is smaller and is less sensitive to outliers. It's also what we use to evaluate investment returns over several years and growth rates.

## The Geometric Mean Return
Let's say you invest \$1,000 in a stock that had a 10\% return in Year 1 and a -10\% return in Year 2. The arithmetic mean shows by year two we'll be back to where we started, that the return is 0\%. Mathematically, $\bar{x} = \frac{0.1 + (-0.1)}{2}=0$ but this ignores compounding! The following table shows that the value of our investment at the end of year two is \$990, a loss of \$10.

In [1]:
# Table.
import pandas as pd
inv = pd.DataFrame({'Return': [0, 0.1, -0.1]})

initial_inv = 1000
inv['val'] = (1 + inv['Return']).cumprod() * initial_inv

inv[1:]

Return     val
1     0.1  1100.0
2    -0.1   990.0

In general, the with multiperiod returns $R_1, R_2, ..., R_n$, the **geometric mean return** $G_R$ is:
$$
G_R = \sqrt[n]{(1 + R_1)\cdot(1 + R_2)\cdot ... \cdot(1 + R_n)} - 1
$$
Thus for our example, the geometric mean is computed as: $G_R = \sqrt[2]{(1+0.1)\cdot(1+(-0.1))} - 1 = -0.005 = -0.5\%$

In [2]:
""" Example 3.12
Calc the geometric mean returns
for Growth and Value.
"""
gv = pd.read_csv('Growth_Value.csv', index_col=0)
gv.head(5)

Growth  Value
Year               
1984   -5.50  -8.59
1985   39.91  22.10
1986   13.03  14.74
1987   -1.70  -8.58
1988   16.05  29.05

In [3]:
# Geometric mean for growth and value.
g_ = (gv['Growth']*.01) + 1
v_ = (gv['Value']*.01) + 1

GR_g = round((g_.product())**(1/len(gv)) - 1, 4)
GR_v = round((v_.product())**(1/len(gv))-1, 4)

print('Growth: ', GR_g*100)
print('Value:', GR_v*100)

Growth:  13.19
Value: 10.35


## Arithmetic Mean vs. Geometric Mean
What about the relevance of the arithmetic mean and the geometric mean for financial returns? They can both be descriptive measures for annual returns. But they have different interpretations in this context. The arithmetic mean allows us to analyze a one-year investment whereas the geometric mean is relevant for multiyear investment analysis. 

For an investment horizon of one year,

In [4]:
round(gv.Growth.mean(), 3)

15.755

is the average annual return for summarizing returns with a one year investment horizon. The geometric mean $13.19$ is the average annual return when the investment horizon is 36 years. As an example, the arithmetic mean return is the relevant metric for an investor who is saving/investing to buy a house in one year's time. The geometric mean return is the relevant measure for an investor saving for retirement.

## The Average Growth Rate
We also use the geometric mean when we calculate an **average growth rate**. For growth rates $g_1, g_2, ..., g_n$ the average growth rate $G_g$ is computed as:
$$
G_g = \sqrt[n]{(1+g_1)\cdot(1+g_2)\cdot ... \cdot(1+g_n)} -1
$$

In [5]:
""" Example 3.13
We have data on sales (in millions of euros)
for a multinational corporation over the past
5 years.
"""
import numpy as np
sales = pd.DataFrame({'Year': range(1,6),
                     'Sales': [13322, 14883, 14203, 14534, 16915]})
sales['Growth_Rate'] = sales['Sales'].pct_change(periods=1)
sales

Year  Sales  Growth_Rate
0     1  13322          NaN
1     2  14883     0.117175
2     3  14203    -0.045690
3     4  14534     0.023305
4     5  16915     0.163823

In [6]:
round(np.prod([1 + x for x in sales['Growth_Rate'] if not np.isnan(x)]) ** (1/(len(sales)-1)) - 1, 4) * 100

6.15

But wait, there's another way to calculate the average growth rate. In this computation, we need not find the growth rates for each year and instead can use the underlying time series values. An alternative formula is:
$$
G_g = \sqrt[n-1]{\frac{x_n}{x_{n-1}} \cdot \frac{x_{n-1}}{x_{n-2}} \cdot ... \cdot \frac{x_2}{x_1}} - 1 = \sqrt[n-1]{\frac{x_n}{x_1}} - 1
$$
where $n-1$ is the number of distinct growth rates. Notice that we only need the first and last value of the time series!

In [7]:
""" Example 3.14
"""
sales

Year  Sales  Growth_Rate
0     1  13322          NaN
1     2  14883     0.117175
2     3  14203    -0.045690
3     4  14534     0.023305
4     5  16915     0.163823

In [8]:
sl = sales.Sales.tolist()
round((sl[-1] / sl[0]) ** (1/(len(sales)-1)) - 1, 4) * 100

6.15